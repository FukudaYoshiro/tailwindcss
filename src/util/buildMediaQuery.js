export default function buildMediaQuery(screens) {
  if (typeof screens === 'string') {
    screens = { min: screens }
  }

  if (!Array.isArray(screens)) {
    screens = [screens]
  }

  return screens
    .map((screen) => {
      if (screen?.hasOwnProperty?.('raw')) {
        return screen.raw
      }

      return Object.entries(screen)
        .map(([feature, value]) => {
          feature = { min: 'min-width', max: 'max-width' }[feature] ?? feature
          return `(${feature}: ${value})`
        })
        .join(' and ')
    })
    .join(', ')
}
