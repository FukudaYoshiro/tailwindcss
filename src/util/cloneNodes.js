export default function cloneNodes(nodes) {
  return nodes.map((node) => node.clone())
}
