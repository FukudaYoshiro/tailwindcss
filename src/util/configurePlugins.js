import _ from 'lodash'

export default function(plugins, pluginConfig, defaultPluginConfig = {}) {
  return Object.keys(plugins)
    .filter(pluginName => {
      return pluginConfig[pluginName] !== false
    })
    .map(pluginName => {
      return plugins[pluginName](_.get(pluginConfig, pluginName, defaultPluginConfig[pluginName]))
    })
}
