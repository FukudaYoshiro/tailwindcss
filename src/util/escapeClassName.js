import parser from 'postcss-selector-parser'
import get from 'lodash/get'

function escapeCommas(className) {
  return className.replace(/\\,/g, '\\2c ')
}

export default function escapeClassName(className) {
  const node = parser.className()
  node.value = className
  return escapeCommas(get(node, 'raws.value', node.value))
}
