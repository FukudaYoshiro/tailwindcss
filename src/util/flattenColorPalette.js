import _ from 'lodash'

export default function flattenColorPalette(colors) {
  const result = _(colors)
    .flatMap((color, name) => {
      if (_.isFunction(color) || !_.isObject(color)) {
        return [[name, color]]
      }

      return _.map(color, (value, key) => {
        const suffix = key === 'DEFAULT' ? '' : `-${key}`
        return [`${name}${suffix}`, value]
      })
    })
    .fromPairs()
    .value()

  return result
}
