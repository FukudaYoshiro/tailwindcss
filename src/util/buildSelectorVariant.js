import escapeClassName from './escapeClassName'
import parser from 'postcss-selector-parser'
import tap from 'lodash/tap'
import get from 'lodash/get'

export default function buildSelectorVariant(selector, variantName, separator, onError = () => {}) {
  return parser(selectors => {
    tap(selectors.first.filter(({ type }) => type === 'class').pop(), classSelector => {
      if (classSelector === undefined) {
        onError('Variant cannot be generated because selector contains no classes.')
        return
      }

      const baseClass = get(classSelector, 'raws.value', classSelector.value)

      classSelector.setPropertyAndEscape(
        'value',
        `${variantName}${escapeClassName(separator)}${baseClass}`
      )
    })
  }).processSync(selector)
}
