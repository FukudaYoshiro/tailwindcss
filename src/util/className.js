import _ from 'lodash'
import escapeClassName from './escapeClassName'

export default function className(base, modifier = 'default') {
  const name = (() => {
    if (modifier === 'default') {
      return base
    }
    return _.startsWith(modifier, '-') ? `-${base}-${modifier.slice(1)}` : `${base}-${modifier}`
  })()

  return `.${escapeClassName(name)}`
}
