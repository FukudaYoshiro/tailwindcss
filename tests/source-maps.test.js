import postcss from 'postcss'
import { parseSourceMaps } from './util/source-maps'
import { crosscheck, runWithSourceMaps as run, html, css, map } from './util/run'

crosscheck(() => {
  it('apply generates source maps', async () => {
    let config = {
      content: [
        {
          raw: html`
            <div class="with-declaration"></div>
            <div class="with-comment"></div>
            <div class="just-apply"></div>
          `,
        },
      ],
      corePlugins: { preflight: false },
    }

    let input = css`
      .with-declaration {
        background-color: red;
        @apply h-4 w-4 bg-green-500;
      }

      .with-comment {
        /* sourcemap will work here too */
        @apply h-4 w-4 bg-red-500;
      }

      .just-apply {
        @apply h-4 w-4 bg-black;
      }
    `

    let result = await run(input, config)
    let { sources, annotations } = parseSourceMaps(result)

    // All CSS generated by Tailwind CSS should be annotated with source maps
    // And always be able to point to the original source file
    expect(sources).not.toContain('<no source>')
    expect(sources.length).toBe(1)

    expect(annotations).toEqual([
      '2:6 -> 2:6',
      '3:8-29 -> 3:8-29',
      '4:8-35 -> 4:8-20',
      '4:8-35 -> 5:8-19',
      '4:8-35 -> 6:8-26',
      '4:8-35 -> 7:8-63',
      '5:6 -> 8:6',
      '7:6 -> 10:6',
      '8:8-41 -> 11:8-41',
      '9:8-33 -> 12:8-20',
      '9:8-33 -> 13:8-19',
      '9:8-33 -> 14:8-26',
      '9:8-33 -> 15:8-63',
      '10:6 -> 16:6',
      '13:8 -> 18:6',
      '13:8-31 -> 19:8-20',
      '13:8-31 -> 20:8-19',
      '13:8-31 -> 21:8-26',
      '13:8 -> 22:8',
      '13:31 -> 23:0',
    ])
  })

  it('preflight + base have source maps', async () => {
    let config = {
      content: [],
    }

    let input = css`
      @tailwind base;
    `

    let result = await run(input, config)
    let { sources, annotations } = parseSourceMaps(result)

    // All CSS generated by Tailwind CSS should be annotated with source maps
    // And always be able to point to the original source file
    expect(sources).not.toContain('<no source>')
    expect(sources.length).toBe(1)

    expect(annotations).toEqual([
      '2:6 -> 1:0',
      '2:20-6 -> 3:1-2',
      '2:20 -> 6:1',
      '2:6 -> 8:0',
      '2:6-20 -> 11:2-32',
      '2:6-20 -> 12:2-25',
      '2:6-20 -> 13:2-29',
      '2:6-20 -> 14:2-31',
      '2:20 -> 15:0',
      '2:6 -> 17:0',
      '2:6-20 -> 19:2-18',
      '2:20 -> 20:0',
      '2:6 -> 22:0',
      '2:20 -> 28:1',
      '2:6 -> 30:0',
      '2:6-20 -> 31:2-26',
      '2:6-20 -> 32:2-40',
      '2:6-20 -> 33:2-26',
      '2:6-20 -> 34:2-21',
      '2:6-20 -> 35:2-230',
      '2:6-20 -> 36:2-39',
      '2:20 -> 37:0',
      '2:6 -> 39:0',
      '2:20 -> 42:1',
      '2:6 -> 44:0',
      '2:6-20 -> 45:2-19',
      '2:6-20 -> 46:2-30',
      '2:20 -> 47:0',
      '2:6 -> 49:0',
      '2:20 -> 53:1',
      '2:6 -> 55:0',
      '2:6-20 -> 56:2-19',
      '2:6-20 -> 57:2-24',
      '2:6-20 -> 58:2-31',
      '2:20 -> 59:0',
      '2:6 -> 61:0',
      '2:20 -> 63:1',
      '2:6 -> 65:0',
      '2:6-20 -> 66:2-35',
      '2:20 -> 67:0',
      '2:6 -> 69:0',
      '2:20 -> 71:1',
      '2:6 -> 73:0',
      '2:6-20 -> 79:2-20',
      '2:6-20 -> 80:2-22',
      '2:20 -> 81:0',
      '2:6 -> 83:0',
      '2:20 -> 85:1',
      '2:6 -> 87:0',
      '2:6-20 -> 88:2-16',
      '2:6-20 -> 89:2-26',
      '2:20 -> 90:0',
      '2:6 -> 92:0',
      '2:20 -> 94:1',
      '2:6 -> 96:0',
      '2:6-20 -> 98:2-21',
      '2:20 -> 99:0',
      '2:6 -> 101:0',
      '2:20 -> 104:1',
      '2:6 -> 106:0',
      '2:6-20 -> 110:2-121',
      '2:6-20 -> 111:2-24',
      '2:20 -> 112:0',
      '2:6 -> 114:0',
      '2:20 -> 116:1',
      '2:6 -> 118:0',
      '2:6-20 -> 119:2-16',
      '2:20 -> 120:0',
      '2:6 -> 122:0',
      '2:20 -> 124:1',
      '2:6 -> 126:0',
      '2:6-20 -> 128:2-16',
      '2:6-20 -> 129:2-16',
      '2:6-20 -> 130:2-20',
      '2:6-20 -> 131:2-26',
      '2:20 -> 132:0',
      '2:6 -> 134:0',
      '2:6-20 -> 135:2-17',
      '2:20 -> 136:0',
      '2:6 -> 138:0',
      '2:6-20 -> 139:2-13',
      '2:20 -> 140:0',
      '2:6 -> 142:0',
      '2:20 -> 146:1',
      '2:6 -> 148:0',
      '2:6-20 -> 149:2-24',
      '2:6-20 -> 150:2-31',
      '2:6-20 -> 151:2-35',
      '2:20 -> 152:0',
      '2:6 -> 154:0',
      '2:20 -> 158:1',
      '2:6 -> 160:0',
      '2:6-20 -> 165:2-30',
      '2:6-20 -> 166:2-25',
      '2:6-20 -> 167:2-30',
      '2:6-20 -> 168:2-30',
      '2:6-20 -> 169:2-24',
      '2:6-20 -> 170:2-19',
      '2:6-20 -> 171:2-20',
      '2:20 -> 172:0',
      '2:6 -> 174:0',
      '2:20 -> 176:1',
      '2:6 -> 178:0',
      '2:6-20 -> 180:2-22',
      '2:20 -> 181:0',
      '2:6 -> 183:0',
      '2:20 -> 186:1',
      '2:6 -> 188:0',
      '2:6-20 -> 192:2-36',
      '2:6-20 -> 193:2-39',
      '2:6-20 -> 194:2-32',
      '2:20 -> 195:0',
      '2:6 -> 197:0',
      '2:20 -> 199:1',
      '2:6 -> 201:0',
      '2:6-20 -> 202:2-15',
      '2:20 -> 203:0',
      '2:6 -> 205:0',
      '2:20 -> 207:1',
      '2:6 -> 209:0',
      '2:6-20 -> 210:2-18',
      '2:20 -> 211:0',
      '2:6 -> 213:0',
      '2:20 -> 215:1',
      '2:6 -> 217:0',
      '2:6-20 -> 218:2-26',
      '2:20 -> 219:0',
      '2:6 -> 221:0',
      '2:20 -> 223:1',
      '2:6 -> 225:0',
      '2:6-20 -> 227:2-14',
      '2:20 -> 228:0',
      '2:6 -> 230:0',
      '2:20 -> 233:1',
      '2:6 -> 235:0',
      '2:6-20 -> 236:2-39',
      '2:6-20 -> 237:2-30',
      '2:20 -> 238:0',
      '2:6 -> 240:0',
      '2:20 -> 242:1',
      '2:6 -> 244:0',
      '2:6-20 -> 245:2-26',
      '2:20 -> 246:0',
      '2:6 -> 248:0',
      '2:20 -> 251:1',
      '2:6 -> 253:0',
      '2:6-20 -> 254:2-36',
      '2:6-20 -> 255:2-23',
      '2:20 -> 256:0',
      '2:6 -> 258:0',
      '2:20 -> 260:1',
      '2:6 -> 262:0',
      '2:6-20 -> 263:2-20',
      '2:20 -> 264:0',
      '2:6 -> 266:0',
      '2:20 -> 268:1',
      '2:6 -> 270:0',
      '2:6-20 -> 283:2-11',
      '2:20 -> 284:0',
      '2:6 -> 286:0',
      '2:6-20 -> 287:2-11',
      '2:6-20 -> 288:2-12',
      '2:20 -> 289:0',
      '2:6 -> 291:0',
      '2:6-20 -> 292:2-12',
      '2:20 -> 293:0',
      '2:6 -> 295:0',
      '2:6-20 -> 298:2-18',
      '2:6-20 -> 299:2-11',
      '2:6-20 -> 300:2-12',
      '2:20 -> 301:0',
      '2:6 -> 303:0',
      '2:20 -> 305:1',
      '2:6 -> 307:0',
      '2:6-20 -> 308:2-18',
      '2:20 -> 309:0',
      '2:6 -> 311:0',
      '2:20 -> 314:1',
      '2:6 -> 316:0',
      '2:6-20 -> 318:2-20',
      '2:6-20 -> 319:2-24',
      '2:20 -> 320:0',
      '2:6 -> 322:0',
      '2:20 -> 324:1',
      '2:6 -> 326:0',
      '2:6-20 -> 328:2-17',
      '2:20 -> 329:0',
      '2:6 -> 331:0',
      '2:20 -> 333:1',
      '2:6 -> 334:0',
      '2:6-20 -> 335:2-17',
      '2:20 -> 336:0',
      '2:6 -> 338:0',
      '2:20 -> 342:1',
      '2:6 -> 344:0',
      '2:6-20 -> 352:2-24',
      '2:6-20 -> 353:2-32',
      '2:20 -> 354:0',
      '2:6 -> 356:0',
      '2:20 -> 358:1',
      '2:6 -> 360:0',
      '2:6-20 -> 362:2-17',
      '2:6-20 -> 363:2-14',
      '2:20 -> 364:0',
      '2:6-20 -> 366:0-72',
      '2:6 -> 367:0',
      '2:6-20 -> 368:2-15',
      '2:20 -> 369:0',
      '2:6 -> 371:0',
      '2:6-20 -> 372:2-26',
      '2:6-20 -> 373:2-26',
      '2:6-20 -> 374:2-21',
      '2:6-20 -> 375:2-21',
      '2:6-20 -> 376:2-16',
      '2:6-20 -> 377:2-16',
      '2:6-20 -> 378:2-16',
      '2:6-20 -> 379:2-17',
      '2:6-20 -> 380:2-17',
      '2:6-20 -> 381:2-15',
      '2:6-20 -> 382:2-15',
      '2:6-20 -> 383:2-20',
      '2:6-20 -> 384:2-40',
      '2:6-20 -> 385:2-17',
      '2:6-20 -> 386:2-22',
      '2:6-20 -> 387:2-24',
      '2:6-20 -> 388:2-25',
      '2:6-20 -> 389:2-26',
      '2:6-20 -> 390:2-20',
      '2:6-20 -> 391:2-29',
      '2:6-20 -> 392:2-30',
      '2:6-20 -> 393:2-40',
      '2:6-20 -> 394:2-36',
      '2:6-20 -> 395:2-29',
      '2:6-20 -> 396:2-24',
      '2:6-20 -> 397:2-32',
      '2:6-20 -> 398:2-14',
      '2:6-20 -> 399:2-20',
      '2:6-20 -> 400:2-18',
      '2:6-20 -> 401:2-19',
      '2:6-20 -> 402:2-20',
      '2:6-20 -> 403:2-16',
      '2:6-20 -> 404:2-18',
      '2:6-20 -> 405:2-15',
      '2:6-20 -> 406:2-21',
      '2:6-20 -> 407:2-23',
      '2:6-20 -> 408:2-29',
      '2:6-20 -> 409:2-27',
      '2:6-20 -> 410:2-28',
      '2:6-20 -> 411:2-29',
      '2:6-20 -> 412:2-25',
      '2:6-20 -> 413:2-26',
      '2:6-20 -> 414:2-27',
      '2:6 -> 415:2',
      '2:20 -> 416:0',
      '2:6 -> 418:0',
      '2:6-20 -> 419:2-26',
      '2:6-20 -> 420:2-26',
      '2:6-20 -> 421:2-21',
      '2:6-20 -> 422:2-21',
      '2:6-20 -> 423:2-16',
      '2:6-20 -> 424:2-16',
      '2:6-20 -> 425:2-16',
      '2:6-20 -> 426:2-17',
      '2:6-20 -> 427:2-17',
      '2:6-20 -> 428:2-15',
      '2:6-20 -> 429:2-15',
      '2:6-20 -> 430:2-20',
      '2:6-20 -> 431:2-40',
      '2:6-20 -> 432:2-17',
      '2:6-20 -> 433:2-22',
      '2:6-20 -> 434:2-24',
      '2:6-20 -> 435:2-25',
      '2:6-20 -> 436:2-26',
      '2:6-20 -> 437:2-20',
      '2:6-20 -> 438:2-29',
      '2:6-20 -> 439:2-30',
      '2:6-20 -> 440:2-40',
      '2:6-20 -> 441:2-36',
      '2:6-20 -> 442:2-29',
      '2:6-20 -> 443:2-24',
      '2:6-20 -> 444:2-32',
      '2:6-20 -> 445:2-14',
      '2:6-20 -> 446:2-20',
      '2:6-20 -> 447:2-18',
      '2:6-20 -> 448:2-19',
      '2:6-20 -> 449:2-20',
      '2:6-20 -> 450:2-16',
      '2:6-20 -> 451:2-18',
      '2:6-20 -> 452:2-15',
      '2:6-20 -> 453:2-21',
      '2:6-20 -> 454:2-23',
      '2:6-20 -> 455:2-29',
      '2:6-20 -> 456:2-27',
      '2:6-20 -> 457:2-28',
      '2:6-20 -> 458:2-29',
      '2:6-20 -> 459:2-25',
      '2:6-20 -> 460:2-26',
      '2:6-20 -> 461:2-27',
      '2:6 -> 462:2',
      '2:20 -> 463:0',
    ])
  })

  test('utilities have source maps', async () => {
    let config = {
      content: [{ raw: `text-red-500` }],
    }

    let input = css`
      @tailwind utilities;
    `

    let result = await run(input, config)
    let { sources, annotations } = parseSourceMaps(result)

    // All CSS generated by Tailwind CSS should be annotated with source maps
    // And always be able to point to the original source file
    expect(sources).not.toContain('<no source>')
    expect(sources.length).toBe(1)

    expect(annotations).toStrictEqual([
      '2:6 -> 1:0',
      '2:6-25 -> 2:4-24',
      '2:6 -> 3:4',
      '2:25 -> 4:0',
    ])
  })

  it('components have source maps', async () => {
    let config = {
      content: [{ raw: `container` }],
    }

    let input = css`
      @tailwind components;
    `

    let result = await run(input, config)
    let { sources, annotations } = parseSourceMaps(result)

    // All CSS generated by Tailwind CSS should be annotated with source maps
    // And always be able to point to the original source file
    expect(sources).not.toContain('<no source>')
    expect(sources.length).toBe(1)

    expect(annotations).toEqual([
      '2:6 -> 1:0',
      '2:6 -> 2:4',
      '2:26 -> 3:0',
      '2:6 -> 4:0',
      '2:6 -> 5:4',
      '2:6 -> 6:8',
      '2:26 -> 7:4',
      '2:26 -> 8:0',
      '2:6 -> 9:0',
      '2:6 -> 10:4',
      '2:6 -> 11:8',
      '2:26 -> 12:4',
      '2:26 -> 13:0',
      '2:6 -> 14:0',
      '2:6 -> 15:4',
      '2:6 -> 16:8',
      '2:26 -> 17:4',
      '2:26 -> 18:0',
      '2:6 -> 19:0',
      '2:6 -> 20:4',
      '2:6 -> 21:8',
      '2:26 -> 22:4',
      '2:26 -> 23:0',
      '2:6 -> 24:0',
      '2:6 -> 25:4',
      '2:6 -> 26:8',
      '2:26 -> 27:4',
      '2:26 -> 28:0',
    ])
  })

  it('source maps for layer rules are not rewritten to point to @tailwind directives', async () => {
    let config = {
      content: [{ raw: `font-normal foo hover:foo` }],
    }

    let utilitiesFile = postcss.parse(
      css`
        @tailwind utilities;
      `,
      { from: 'components.css', map: { prev: map } }
    )

    let mainCssFile = postcss.parse(
      css`
        @layer utilities {
          .foo {
            background-color: red;
          }
        }
      `,
      { from: 'input.css', map: { prev: map } }
    )

    // Just pretend that there's an @import in `mainCssFile` that imports the nodes from `utilitiesFile`
    let input = postcss.root({
      nodes: [...utilitiesFile.nodes, ...mainCssFile.nodes],
      source: mainCssFile.source,
    })

    let result = await run(input, config)

    let { sources, annotations } = parseSourceMaps(result)

    // All CSS generated by Tailwind CSS should be annotated with source maps
    // And always be able to point to the original source file
    expect(sources).not.toContain('<no source>')

    // And we should see that the source map for the layer rule is not rewritten
    // to point to the @tailwind directive but instead points to the original
    expect(sources.length).toBe(2)
    expect(sources).toEqual(['components.css', 'input.css'])

    expect(annotations).toEqual([
      '2:8 -> 1:0',
      '2:8 -> 2:12',
      '2:27 -> 3:0',
      '3:10 -> 4:10',
      '4:12-33 -> 5:12-33',
      '5:10 -> 6:10',
      '3:10 -> 7:10',
      '4:12-33 -> 8:12-33',
      '5:10 -> 9:10',
    ])
  })
})
