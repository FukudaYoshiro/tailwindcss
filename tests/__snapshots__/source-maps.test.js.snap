// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply generates source maps 1`] = `
Array [
  "2:4 -> 2:4",
  "3:6-27 -> 3:6-27",
  "4:6-33 -> 4:6-18",
  "4:6-33 -> 5:6-17",
  "4:6-33 -> 6:6-24",
  "4:6-33 -> 7:6-61",
  "5:4 -> 8:4",
  "7:4 -> 10:4",
  "8:6-39 -> 11:6-39",
  "9:6-31 -> 12:6-18",
  "9:6-31 -> 13:6-17",
  "9:6-31 -> 14:6-24",
  "9:6-31 -> 15:6-61",
  "10:4 -> 16:4",
  "13:6 -> 18:4",
  "13:6-29 -> 19:6-18",
  "13:6-29 -> 20:6-17",
  "13:6-29 -> 21:6-24",
  "13:6 -> 22:6",
  "13:29 -> 23:0",
]
`;

exports[`components have source maps 1`] = `
Array [
  "2:4 -> 1:0",
  "2:4 -> 2:4",
  "2:24 -> 3:0",
  "2:4 -> 4:0",
  "2:4 -> 5:4",
  "2:4 -> 6:8",
  "2:24 -> 7:4",
  "2:24 -> 8:0",
  "2:4 -> 9:0",
  "2:4 -> 10:4",
  "2:4 -> 11:8",
  "2:24 -> 12:4",
  "2:24 -> 13:0",
  "2:4 -> 14:0",
  "2:4 -> 15:4",
  "2:4 -> 16:8",
  "2:24 -> 17:4",
  "2:24 -> 18:0",
  "2:4 -> 19:0",
  "2:4 -> 20:4",
  "2:4 -> 21:8",
  "2:24 -> 22:4",
  "2:24 -> 23:0",
  "2:4 -> 24:0",
  "2:4 -> 25:4",
  "2:4 -> 26:8",
  "2:24 -> 27:4",
  "2:24 -> 28:0",
]
`;

exports[`preflight + base have source maps 1`] = `
Array [
  "2:4 -> 1:0",
  "2:18-4 -> 3:1-2",
  "2:18 -> 6:1",
  "2:4 -> 8:0",
  "2:4-18 -> 11:2-32",
  "2:4-18 -> 12:2-25",
  "2:4-18 -> 13:2-29",
  "2:4-18 -> 14:2-31",
  "2:18 -> 15:0",
  "2:4 -> 17:0",
  "2:4-18 -> 19:2-18",
  "2:18 -> 20:0",
  "2:4 -> 22:0",
  "2:18 -> 27:1",
  "2:4 -> 29:0",
  "2:4-18 -> 30:2-26",
  "2:4-18 -> 31:2-40",
  "2:4-18 -> 32:2-26",
  "2:4-18 -> 33:2-21",
  "2:4-18 -> 34:2-230",
  "2:18 -> 35:0",
  "2:4 -> 37:0",
  "2:18 -> 40:1",
  "2:4 -> 42:0",
  "2:4-18 -> 43:2-19",
  "2:4-18 -> 44:2-30",
  "2:18 -> 45:0",
  "2:4 -> 47:0",
  "2:18 -> 51:1",
  "2:4 -> 53:0",
  "2:4-18 -> 54:2-19",
  "2:4-18 -> 55:2-24",
  "2:4-18 -> 56:2-31",
  "2:18 -> 57:0",
  "2:4 -> 59:0",
  "2:18 -> 61:1",
  "2:4 -> 63:0",
  "2:4-18 -> 64:2-35",
  "2:18 -> 65:0",
  "2:4 -> 67:0",
  "2:18 -> 69:1",
  "2:4 -> 71:0",
  "2:4-18 -> 77:2-20",
  "2:4-18 -> 78:2-22",
  "2:18 -> 79:0",
  "2:4 -> 81:0",
  "2:18 -> 83:1",
  "2:4 -> 85:0",
  "2:4-18 -> 86:2-16",
  "2:4-18 -> 87:2-26",
  "2:18 -> 88:0",
  "2:4 -> 90:0",
  "2:18 -> 92:1",
  "2:4 -> 94:0",
  "2:4-18 -> 96:2-21",
  "2:18 -> 97:0",
  "2:4 -> 99:0",
  "2:18 -> 102:1",
  "2:4 -> 104:0",
  "2:4-18 -> 108:2-121",
  "2:4-18 -> 109:2-24",
  "2:18 -> 110:0",
  "2:4 -> 112:0",
  "2:18 -> 114:1",
  "2:4 -> 116:0",
  "2:4-18 -> 117:2-16",
  "2:18 -> 118:0",
  "2:4 -> 120:0",
  "2:18 -> 122:1",
  "2:4 -> 124:0",
  "2:4-18 -> 126:2-16",
  "2:4-18 -> 127:2-16",
  "2:4-18 -> 128:2-20",
  "2:4-18 -> 129:2-26",
  "2:18 -> 130:0",
  "2:4 -> 132:0",
  "2:4-18 -> 133:2-17",
  "2:18 -> 134:0",
  "2:4 -> 136:0",
  "2:4-18 -> 137:2-13",
  "2:18 -> 138:0",
  "2:4 -> 140:0",
  "2:18 -> 144:1",
  "2:4 -> 146:0",
  "2:4-18 -> 147:2-24",
  "2:4-18 -> 148:2-31",
  "2:4-18 -> 149:2-35",
  "2:18 -> 150:0",
  "2:4 -> 152:0",
  "2:18 -> 156:1",
  "2:4 -> 158:0",
  "2:4-18 -> 163:2-30",
  "2:4-18 -> 164:2-25",
  "2:4-18 -> 165:2-30",
  "2:4-18 -> 166:2-30",
  "2:4-18 -> 167:2-24",
  "2:4-18 -> 168:2-19",
  "2:4-18 -> 169:2-20",
  "2:18 -> 170:0",
  "2:4 -> 172:0",
  "2:18 -> 174:1",
  "2:4 -> 176:0",
  "2:4-18 -> 178:2-22",
  "2:18 -> 179:0",
  "2:4 -> 181:0",
  "2:18 -> 184:1",
  "2:4 -> 186:0",
  "2:4-18 -> 190:2-36",
  "2:4-18 -> 191:2-39",
  "2:4-18 -> 192:2-32",
  "2:18 -> 193:0",
  "2:4 -> 195:0",
  "2:18 -> 197:1",
  "2:4 -> 199:0",
  "2:4-18 -> 200:2-15",
  "2:18 -> 201:0",
  "2:4 -> 203:0",
  "2:18 -> 205:1",
  "2:4 -> 207:0",
  "2:4-18 -> 208:2-18",
  "2:18 -> 209:0",
  "2:4 -> 211:0",
  "2:18 -> 213:1",
  "2:4 -> 215:0",
  "2:4-18 -> 216:2-26",
  "2:18 -> 217:0",
  "2:4 -> 219:0",
  "2:18 -> 221:1",
  "2:4 -> 223:0",
  "2:4-18 -> 225:2-14",
  "2:18 -> 226:0",
  "2:4 -> 228:0",
  "2:18 -> 231:1",
  "2:4 -> 233:0",
  "2:4-18 -> 234:2-39",
  "2:4-18 -> 235:2-30",
  "2:18 -> 236:0",
  "2:4 -> 238:0",
  "2:18 -> 240:1",
  "2:4 -> 242:0",
  "2:4-18 -> 243:2-26",
  "2:18 -> 244:0",
  "2:4 -> 246:0",
  "2:18 -> 249:1",
  "2:4 -> 251:0",
  "2:4-18 -> 252:2-36",
  "2:4-18 -> 253:2-23",
  "2:18 -> 254:0",
  "2:4 -> 256:0",
  "2:18 -> 258:1",
  "2:4 -> 260:0",
  "2:4-18 -> 261:2-20",
  "2:18 -> 262:0",
  "2:4 -> 264:0",
  "2:18 -> 266:1",
  "2:4 -> 268:0",
  "2:4-18 -> 281:2-11",
  "2:18 -> 282:0",
  "2:4 -> 284:0",
  "2:4-18 -> 285:2-11",
  "2:4-18 -> 286:2-12",
  "2:18 -> 287:0",
  "2:4 -> 289:0",
  "2:4-18 -> 290:2-12",
  "2:18 -> 291:0",
  "2:4 -> 293:0",
  "2:4-18 -> 296:2-18",
  "2:4-18 -> 297:2-11",
  "2:4-18 -> 298:2-12",
  "2:18 -> 299:0",
  "2:4 -> 301:0",
  "2:18 -> 303:1",
  "2:4 -> 305:0",
  "2:4-18 -> 306:2-18",
  "2:18 -> 307:0",
  "2:4 -> 309:0",
  "2:18 -> 312:1",
  "2:4 -> 314:0",
  "2:4-18 -> 316:2-20",
  "2:4-18 -> 317:2-24",
  "2:18 -> 318:0",
  "2:4 -> 320:0",
  "2:18 -> 324:1",
  "2:4 -> 326:0",
  "2:4-18 -> 334:2-24",
  "2:4-18 -> 335:2-32",
  "2:18 -> 336:0",
  "2:4 -> 338:0",
  "2:18 -> 340:1",
  "2:4 -> 342:0",
  "2:4-18 -> 344:2-17",
  "2:4-18 -> 345:2-14",
  "2:18 -> 346:0",
  "2:4 -> 348:0",
  "2:4-18 -> 349:2-26",
  "2:4-18 -> 350:2-26",
  "2:4-18 -> 351:2-21",
  "2:4-18 -> 352:2-21",
  "2:4-18 -> 353:2-16",
  "2:4-18 -> 354:2-16",
  "2:4-18 -> 355:2-16",
  "2:4-18 -> 356:2-17",
  "2:4-18 -> 357:2-17",
  "2:4-18 -> 358:2-15",
  "2:4-18 -> 359:2-15",
  "2:4-18 -> 360:2-20",
  "2:4-18 -> 361:2-40",
  "2:4-18 -> 362:2-17",
  "2:4-18 -> 363:2-22",
  "2:4-18 -> 364:2-24",
  "2:4-18 -> 365:2-25",
  "2:4-18 -> 366:2-26",
  "2:4-18 -> 367:2-20",
  "2:4-18 -> 368:2-29",
  "2:4-18 -> 369:2-30",
  "2:4-18 -> 370:2-40",
  "2:4-18 -> 371:2-36",
  "2:4-18 -> 372:2-29",
  "2:4-18 -> 373:2-24",
  "2:4-18 -> 374:2-32",
  "2:4-18 -> 375:2-14",
  "2:4-18 -> 376:2-20",
  "2:4-18 -> 377:2-18",
  "2:4-18 -> 378:2-19",
  "2:4-18 -> 379:2-20",
  "2:4-18 -> 380:2-16",
  "2:4-18 -> 381:2-18",
  "2:4-18 -> 382:2-15",
  "2:4-18 -> 383:2-21",
  "2:4-18 -> 384:2-23",
  "2:4-18 -> 385:2-29",
  "2:4-18 -> 386:2-27",
  "2:4-18 -> 387:2-28",
  "2:4-18 -> 388:2-29",
  "2:4-18 -> 389:2-25",
  "2:4-18 -> 390:2-26",
  "2:4-18 -> 391:2-27",
  "2:4 -> 392:2",
  "2:18 -> 393:0",
  "2:4 -> 395:0",
  "2:4-18 -> 396:2-26",
  "2:4-18 -> 397:2-26",
  "2:4-18 -> 398:2-21",
  "2:4-18 -> 399:2-21",
  "2:4-18 -> 400:2-16",
  "2:4-18 -> 401:2-16",
  "2:4-18 -> 402:2-16",
  "2:4-18 -> 403:2-17",
  "2:4-18 -> 404:2-17",
  "2:4-18 -> 405:2-15",
  "2:4-18 -> 406:2-15",
  "2:4-18 -> 407:2-20",
  "2:4-18 -> 408:2-40",
  "2:4-18 -> 409:2-17",
  "2:4-18 -> 410:2-22",
  "2:4-18 -> 411:2-24",
  "2:4-18 -> 412:2-25",
  "2:4-18 -> 413:2-26",
  "2:4-18 -> 414:2-20",
  "2:4-18 -> 415:2-29",
  "2:4-18 -> 416:2-30",
  "2:4-18 -> 417:2-40",
  "2:4-18 -> 418:2-36",
  "2:4-18 -> 419:2-29",
  "2:4-18 -> 420:2-24",
  "2:4-18 -> 421:2-32",
  "2:4-18 -> 422:2-14",
  "2:4-18 -> 423:2-20",
  "2:4-18 -> 424:2-18",
  "2:4-18 -> 425:2-19",
  "2:4-18 -> 426:2-20",
  "2:4-18 -> 427:2-16",
  "2:4-18 -> 428:2-18",
  "2:4-18 -> 429:2-15",
  "2:4-18 -> 430:2-21",
  "2:4-18 -> 431:2-23",
  "2:4-18 -> 432:2-29",
  "2:4-18 -> 433:2-27",
  "2:4-18 -> 434:2-28",
  "2:4-18 -> 435:2-29",
  "2:4-18 -> 436:2-25",
  "2:4-18 -> 437:2-26",
  "2:4-18 -> 438:2-27",
  "2:4 -> 439:2",
  "2:18 -> 440:0",
]
`;

exports[`source maps for layer rules are not rewritten to point to @tailwind directives 1`] = `
Array [
  "2:6 -> 1:0",
  "2:6 -> 2:10",
  "2:25 -> 3:0",
  "3:8 -> 4:8",
  "4:10-31 -> 5:10-31",
  "5:8 -> 6:8",
  "3:8 -> 7:8",
  "4:10-31 -> 8:10-31",
  "5:8 -> 9:8",
]
`;
